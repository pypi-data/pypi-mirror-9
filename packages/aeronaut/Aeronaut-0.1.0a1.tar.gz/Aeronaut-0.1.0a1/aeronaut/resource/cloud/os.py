from aeronaut.resource.cloud.resource import Resource


class OperatingSystem(Resource):

    def _root_(self):
        return "operatingSystem"

    def _members_(self):
        return {
            "id": {
                "xpath": "./@*[local-name()='id']"},

            "display_name": {
                "xpath": "./@*[local-name()='displayName']"},

            "type": {
                "xpath": "./@*[local-name()='type']"},
        }
