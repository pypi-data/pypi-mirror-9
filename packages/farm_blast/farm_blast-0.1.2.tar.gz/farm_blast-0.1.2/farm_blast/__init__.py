__all__ = ['utils', 'blast', 'pipeline']
from farm_blast import *
