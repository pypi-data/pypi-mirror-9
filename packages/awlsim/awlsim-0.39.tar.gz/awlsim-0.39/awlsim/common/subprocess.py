# -*- coding: utf-8 -*-
#
# subprocess wrapper
#
# Copyright 2014 Michael Buesch <m@bues.ch>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#

from __future__ import division, absolute_import, print_function, unicode_literals
from awlsim.common.compat import *

import distutils.spawn

if isIronPython:
	import os
	import signal
else:
	import subprocess

def findExecutable(executable):
	return distutils.spawn.find_executable(executable)

class PopenWrapper(object):
	def __init__(self, argv, env):
		if isIronPython:
			self.__pid = os.spawnve(os.P_NOWAIT, argv[0], argv, env)
		else:
			self.__proc = subprocess.Popen(argv, env = env, shell = False)

	def terminate(self):
		if isIronPython:
			try:
				os.kill(self.__pid, signal.SIGTERM)
			except ValueError:
				pass
		else:
			self.__proc.terminate()

	def wait(self):
		if isIronPython:
			pass#TODO
		else:
			self.__proc.wait()
