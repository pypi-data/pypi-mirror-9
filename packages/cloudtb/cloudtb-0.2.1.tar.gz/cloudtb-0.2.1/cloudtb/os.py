# -*- coding: utf-8 -*-
from __future__ import (absolute_import, division, print_function,
                        unicode_literals)
from os import path as _path


def abspath(path):
    '''A better abspath. Automatically expands user'''
    return _path.abspath(_path.expanduser(path))


def split(path):
    '''A better split function. Completely splits a path'''
    _split = _path.split
    folders = []
    while 1:
        path, folder = _split(path)
        if folder:
            folders.append(folder)
        else:
            if path:
                folders.append(path)
            break

    folders.reverse()
    return folders
