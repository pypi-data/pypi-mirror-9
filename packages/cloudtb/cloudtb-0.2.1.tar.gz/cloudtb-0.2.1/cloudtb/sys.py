# -*- coding: utf-8 -*-
from __future__ import (absolute_import, division, print_function,
                        unicode_literals)

import os
import sys


def supports_color():
    """Returns True if the running system's terminal supports color,
    and False otherwise.
    based on django.core.management.color
    """
    plat = sys.platform
    if plat == 'Pocket PC' or (plat == 'win32' and 'ANSICON' not in os.environ):
        return False
    if not hasattr(sys.stdout, 'isatty') or not sys.stdout.isatty():
        return False
    return True


hascolor = supports_color()  # fast access
