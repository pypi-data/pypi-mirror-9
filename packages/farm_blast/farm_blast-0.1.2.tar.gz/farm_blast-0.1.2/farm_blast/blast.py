import os

class Error (Exception): pass

ref_not_protein_types = set([
    'blastn',
    'blastn-short',
    'dc-megablast',
    'megablast',
    'rmblastn',
    'tblastn',
    'tblastx',
])

ref_protein_types = set([
    'blastp',
    'blastp-short',
    'deltablast',
    'blastx'
])

class Blast:
    def __init__(
         self,
         reference,
         query,
         outfile='blast.out',
         blastall=False,
         blast_type='blastn',
         evalue=None,
         word_size=None,
         no_filter=False,
         extra_options=''):

        self.reference = reference
        self.query = query
        self.outfile = outfile
        self.blastall = blastall
        self.blast_type = blast_type
        self.evalue = evalue
        self.word_size = word_size
        self.no_filter = no_filter
        self.extra_options = extra_options

        if self.blast_type in ref_not_protein_types:
            self.protein_reference = False
        elif self.blast_type in ref_protein_types:
            self.protein_reference = True
        else:
            raise Error('blast_type must be one of ' + ' '.join(list(ref_not_protein_types) + list(ref_protein_types)))

    def blast_db_exists(self):
        if self.protein_reference:
            test_files1 = [self.reference + '.' + e for e in ['phr', 'pin', 'psq']]
            test_files2 = [self.reference + '.00.' + e for e in ['phr', 'pin', 'psq']]
        else:
            test_files1 = [self.reference + '.' + e for e in ['nhr', 'nin', 'nsq']]
            test_files2 = [self.reference + '.00.' + e for e in ['nhr', 'nin', 'nsq']]

        if False not in [os.path.isfile(fname) for fname in test_files1] \
             or False not in [os.path.isfile(fname) for fname in test_files2]:
            return True

        return False
         

    def format_database_command(self):
        if self.blast_db_exists():
            return None

        if self.protein_reference:
            dbtype = 'prot'
            popt = 'T'
        else:
            dbtype = 'nucl'
            popt = 'F'

        if self.blastall:
            return ' '.join(['formatdb -p', popt, '-i', self.reference])
        else:
            return ' '.join(['makeblastdb -dbtype', dbtype, '-in', self.reference])


    def _make_options_string(self):
        opts = []

        if self.blastall:
            opts.extend(['-m', '8'])

            if self.evalue:
                opts.extend(['-e', str(self.evalue)])
            if self.word_size:
                opts.extend(['-W', str(self.word_size)])
            if self.no_filter:
                opts.extend(['-F', 'F'])
        else:
            opts.extend(['-outfmt', '6'])

            if self.evalue:
                opts.extend(['-evalue', str(self.evalue)])
            if self.word_size:
                opts.extend(['-word_size', str(self.word_size)])
            if self.no_filter and not self.protein_reference:
                opts.extend(['-dust', 'no'])
            if self.protein_reference and not self.no_filter:
                opts.extend(['-seg', 'yes'])

        if self.extra_options:
            opts.append(self.extra_options)

        return ' '.join(opts)


    def _make_io_string(self):
        if self.blastall:
            opts = [
                '-d', self.reference,
                '-i', self.query,
                '-o', self.outfile
            ]
        else:
            opts = [
                '-db', self.reference,
                '-query', self.query,
                '-out', self.outfile
            ]

        return ' '.join(opts)
        

    def _make_blast_type_string(self):
        if self.blastall:
            if self.blast_type == 'megablast':
                return 'blastall -p blastn -n T'
            else:
                return 'blastall -p ' + self.blast_type
        else:
            type_to_string = {
                'blastn': 'blastn -task blastn',
                'blastn-short': 'blastn -task blastn-short',
                'dc-megablast': 'blastn -task dc-megablast',
                'megablast': 'blastn -task megablast',
                'rmblastn': 'blastn -task rmblastn',
                'tblastn': 'tblastn',
                'tblastx': 'tblastx',
                'blastp': 'blastp -task blastp',
                'blastp-short': 'blastp -task blastp-short',
                'deltablast': 'blastp -task deltablast',
                'blastx': 'blastx'
            }

            try:
                return type_to_string[self.blast_type]
            except:
                raise Error('Problem with blast type: ' + self.blast_type)


    def get_run_command(self):
        return ' '.join([
            self._make_blast_type_string(),
            self._make_io_string(),
            self._make_options_string()
        ])
