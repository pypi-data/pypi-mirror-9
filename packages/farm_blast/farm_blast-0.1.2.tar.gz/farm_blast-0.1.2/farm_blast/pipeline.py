import os
import sys
import argparse
import time
import glob
from farmpy import lsf
from farm_blast import blast, utils

class Error (Exception): pass

parser = argparse.ArgumentParser(
    description = 'Run BLAST in parallel on the farm',
    usage = '%(prog)s [options] <reference> <query>')

parser.add_argument('--no_bsub', action='store_true', help=argparse.SUPPRESS)
parser.add_argument('--fix_coords_in_blast_output', action='store_true', help=argparse.SUPPRESS)
parser.add_argument('--split_bases_tolerance', type=int, default=1000, help=argparse.SUPPRESS)
parser.add_argument('--test', action='store_true', help=argparse.SUPPRESS)

common_blast_group = parser.add_argument_group('Common BLAST options')
common_blast_group.add_argument('--blastall', action='store_true', help='Use blastall instead of the default blast+')
common_blast_group.add_argument('-p', '--blast_type', help='Type of blast to run [%(default)s]', choices=sorted(list(blast.ref_not_protein_types)) + sorted(list(blast.ref_protein_types)), default='blastn')
common_blast_group.add_argument('-e', '--evalue', help='Set the evalue cutoff')
common_blast_group.add_argument('-W', '--word_size', help='Set the word size')
common_blast_group.add_argument('--no_filter', action='store_true', help='Do not filter query sequence (equivalent to -F F in blastall, -dust no in blast+). By default, the query will be filtered')

bsub_group = parser.add_argument_group('Bsub options')
bsub_group.add_argument('-q', '--bsub_queue', help='Queue in which all jobs are run [%(default)s]', default = 'normal', metavar='Queue_name')
bsub_group.add_argument('--blast_mem', type=float, help='Memory limit in GB for the farm jobs that run BLAST. Default is 0.5, except set to 5 if blastall tblastx is used. Defaults doubled if --no_filter used', metavar='FLOAT', default=None)
bsub_group.add_argument('--bsub_name_prefix', help='Set the prefix of the names of the bsub jobs', default=None)


advanced_opts_group = parser.add_argument_group('Advanced options')
advanced_opts_group.add_argument('--act', action='store_true', help='Make ACT-friendly blast file, by concatenating all reference sequences together and all query sequences together before blasting.')
advanced_opts_group.add_argument('--blast_options', help='Put any extra options to the blast call (i.e. blastall, blastn, blastx ...etc) in quotes. e.g. --blast_options "-r 2". Whatever you put in here is NOT sanity checked.', default = '', metavar='"options in quotes"')
advanced_opts_group.add_argument('--debug', action='store_true', help='Just make scripts etc but do not run anything')
advanced_opts_group.add_argument('--outdir', help='Name of output directory (must not exist already)', metavar='output directory', default=None)
advanced_opts_group.add_argument('--split_bases', type=int, help='Number of bases in each split file of query. Default is 500000, except set to 200000 if blastall tblastx is used', metavar='INT', default=None)

parser.add_argument('reference', help='Name of reference file. Does not need to be indexed already. If not indexed, can be any format from FASTA, FASTQ, GFF3, EMBL, Phylip, GBK', metavar='reference')
parser.add_argument('query', help='Name of query file. Can be any format from FASTA, FASTQ, GFF3, EMBL, Phylip, GBK', metavar='query')

def get_opts(args=None):
    return parser.parse_args(args=args)


class Pipeline:
    def __init__(self, options, farm_blast_script):
        if options.outdir is None:
            if options.blastall:
                version = 'blastall'
            else:
                version = 'blast_plus'

            options.outdir = '.'.join(['Farm_blast', os.path.basename(options.reference), os.path.basename(options.query), version, options.blast_type, 'out'])

        self.outdir = os.path.abspath(options.outdir)
        self.reference = os.path.abspath(options.reference)
        self.query = os.path.abspath(options.query)
        self.bsub_queue = options.bsub_queue
        self.farm_blast_script = farm_blast_script
        self.test = options.test
        self.union_for_act = options.act

        self.blast = blast.Blast(
            self.reference,
            'query.split.INDEX',
            outfile='tmp.array.out.INDEX',
            blastall=options.blastall,
            blast_type=options.blast_type,
            evalue=options.evalue,
            word_size=options.word_size,
            no_filter=options.no_filter,
            extra_options=options.blast_options
        )

        self.setup_script = '01.setup.sh'
        self.start_array_script = '02.run_array.sh'
        self.combine_script = '03.combine.sh'
        if options.bsub_name_prefix is None:
            self.bsub_name_prefix = 'farm_blast:' + self.outdir
        else:
            self.bsub_name_prefix = options.bsub_name_prefix

        if options.no_bsub:
            self.no_bsub = True
            self.memory_units = 'MB'
        else:
            self.no_bsub = False
            self.memory_units = None

        self.debug = options.debug
        self.split_bases_tolerance = options.split_bases_tolerance

        self.files_to_delete = [
            'tmp.array.*',
            'query.split.*',
            'blast.out.tmp.gz',
            '02.array.id',
            '03.combine.sh.id',
        ]

        if not options.blast_mem:
            if self.blast.blastall and self.blast.blast_type == 'tblastx':
                self.array_mem = 5
            else:
                self.array_mem = 0.5

            if self.blast.no_filter:
                self.array_mem *= 2
        else:
            self.array_mem = options.blast_mem

        if not options.split_bases:
            if self.blast.blastall and self.blast.blast_type == 'tblastx':
                self.split_bases = 200000
            else:
                self.split_bases = 500000
        else:
            self.split_bases = options.split_bases


    def _make_setup_script(self, script_name=None):
        if script_name is None:
            script_name = self.setup_script
        try:
            f = open(script_name, 'w')
        except:
            raise Error('Error opening setup script "' + script_name + '" for writing')

        print('set -e', file=f)
        if not self.blast.blast_db_exists() or self.union_for_act:
            if self.union_for_act:
                print('fastaq merge', self.reference, '- |',
                      'fastaq to_fasta -s - reference.fa', file=f)
            else:
                print('fastaq to_fasta -s', self.reference, 'reference.fa', file=f)
            self.reference = 'reference.fa'
            self.blast.reference = self.reference
            print(self.blast.format_database_command(), file=f)
            self.files_to_delete.append('reference.*')

        # blast strips off everything after the first whitespace, so do this
        # before chunking so names stay consistent with query fasta and in blast output
        if self.union_for_act:
            print('fastaq merge', self.query, '- |',
                  'fastaq to_fasta -s - - |', end=' ', file=f)
        else:
            print('fastaq to_fasta -s', self.query, '- |', end=' ', file=f)

        print('fastaq chunker --skip_all_Ns', '-', 'query.split', self.split_bases, self.split_bases_tolerance, file=f)
        f.close()


    def _make_setup_job(self):
        self.setup_job = lsf.Job(
            self.setup_script + '.o',
            self.setup_script + '.e',
            self.bsub_name_prefix + '.setup',
            self.bsub_queue,
            1,
            'bash ' + self.setup_script,
            memory_units=self.memory_units,
        )


    def _make_array_job(self):
        self.array_job = lsf.Job(
            'tmp.array.o',
            'tmp.array.e',
            self.bsub_name_prefix + '.array',
            self.bsub_queue,
            self.array_mem,
            self.blast.get_run_command(),
            array_start=1,
            array_end=r'''$n''',
            memory_units=self.memory_units,
            max_array_size=100
        )


    def _make_start_array_job(self):
        self.start_array_job = lsf.Job(
            self.start_array_script + '.o',
            self.start_array_script + '.e',
            self.bsub_name_prefix + '.start_array',
            'small',
            0.1,
            'bash ' + self.start_array_script,
            no_resources=True  # we have to do this otherwise bmod fails! LSF bug?
        )


    def _make_start_array_script(self, script_name=None):
        if script_name is None:
            script_name = self.start_array_script
        try:
            f = open(script_name, 'w')
        except:
            raise Error('Error writing script "' + script_name + '"')
        print('set -e', file=f)
        print(r'''n=`ls query.split.* | grep -v coords | wc -l`''', file=f)
        print(str(self.array_job) + r''' |  awk '{print substr($2,2,length($2)-2)}' > 02.array.id''', file=f)
        print(r'''array_id=`cat 02.array.id`
combine_id=`cat ''' + self.combine_script + r'''.id`
bmod -w "done($array_id)" $combine_id''', file=f)
        f.close()


    def _make_combine_job(self):
        self.combine_job = lsf.Job(
            self.combine_script + '.o',
            self.combine_script + '.e',
            self.bsub_name_prefix + '.combine',
            self.bsub_queue,
            0.5,
            'bash ' + self.combine_script,
            memory_units=self.memory_units,
            threads=2
        )


    def _make_combine_script(self, script_name=None):
        if script_name is None:
            script_name = self.combine_script
        try:
            f = open(script_name, 'w')
        except:
            raise Error('Error writing script "' + script_name + '"')

        if not self.no_bsub:
            print('set -e', file=f)

        print(r'''cat tmp.array.e.* > 02.array.e
cat tmp.array.o.* > 02.array.o
cat tmp.array.out.* | gzip -9 -c > blast.out.tmp.gz''', file=f)
        if self.test:
            p = os.path.dirname(self.farm_blast_script)
            p = os.path.join(p, os.pardir)
            p = os.path.normpath(p)
            print('PYTHONPATH=' + p + ':$PYTHONPATH', file=f)
        if self.test:
            print(self.farm_blast_script, '--test --fix_coords_in_blast_output x x', file=f)
        else:
            print(self.farm_blast_script, '--fix_coords_in_blast_output x x', file=f)
        print('rm', ' '.join(self.files_to_delete), file=f)
        print('touch FINISHED', file=f)
        f.close()


    def run(self):
        '''Runs the whole blast pipeline'''
        # Here's the fun part: the first job splits the query file into
        # chunks. Don't know the number of chunks until that job has finished.
        # Therefore don't know the size of the job array until the chunking
        # job has finished.
        #
        # Here's what's going to happen:
        # 1. Submit chunking job.
        # 2. Submit job to submit an array, to run when Job1 finishes.
        #      - this is NOT the array itself!
        # 3. Submit job that combines the results of the job array and tidies
        # up files. It is initially set to run when the job that starts
        # the array finishes.
        # 4. When the job that starts the array actually runs, it does this:
        #       - figures out the size of the job array and submits it
        #       - changes the dependency of the last combine job, so that
        #         the combien job runs when the array has finished
        # 5. When the array finishes, Job3 will then run.

        try:
            os.mkdir(self.outdir)
        except:
            print('Error making output directory', self.outdir, file=sys.stderr)
            sys.exit(1)

        original_dir = os.getcwd()
        os.chdir(self.outdir)
        self._make_setup_script()
        self._make_setup_job()
        self._make_array_job()
        self._make_start_array_script()
        self._make_start_array_job()
        self._make_combine_script()
        self._make_combine_job()

        if self.debug:
            sys.exit()

        if self.no_bsub:
            self.setup_job.run_not_bsubbed()
            # get size of job array
            files_count = len(glob.glob('query.split.*')) - 1
            self.array_job.array_end = files_count
            print(self.array_job)
            print(self.array_job.array_start)
            print(self.array_job.array_end)
            print(self.array_job._make_command_string().replace('\$LSB_JOBINDEX', '$LSB_JOBINDEX'))
            self.array_job.run_not_bsubbed()

            # a little hack here to make the farm_blast script run
            this_script = os.path.realpath(__file__)
            this_script_dir = os.path.dirname(this_script)
            module_root_dir = os.path.join(this_script_dir, '..')
            module_root_dir = os.path.normpath(module_root_dir)
            os.environ["PATH"] = os.path.join(module_root_dir, 'scripts:') + os.environ["PATH"]
            os.environ["PYTHONPATH"] = module_root_dir + ':' + os.environ["PYTHONPATH"]
            self.combine_job.run_not_bsubbed()
        else:
            self.setup_job.run()
            time.sleep(1)
            self.start_array_job.add_dependency(self.setup_job.job_id)
            self.start_array_job.run()
            time.sleep(1)
            self.combine_job.add_dependency(self.start_array_job.job_id)
            self.combine_job.run()
            time.sleep(1)
            try:
                f = open(self.combine_script + '.id', 'w')
            except:
                raise Error('Error opening file "' + self.combine_script + '.id' + '" for writing')
            print(self.combine_job.job_id, file=f)
            f.close()
            print('Jobs submitted to the farm.')
            print('Final job id is', self.combine_job.job_id)
            print('\nPipeline finished OK when this file is written:\n   ', os.path.join(self.outdir, 'FINISHED'))
            print('\nFinal file will be called:\n   ', os.path.join(self.outdir, 'blast.out.gz'))

        os.chdir(original_dir)
