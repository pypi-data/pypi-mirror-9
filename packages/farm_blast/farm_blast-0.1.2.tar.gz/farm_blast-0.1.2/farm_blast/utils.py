from pyfastaq import utils
import sys

class Error (Exception): pass


def offset_coords_file_to_dict(filename):
    f = utils.open_file_read(filename)
    offsets = {}

    for line in f:
        (seq, ref, offset) = line.rstrip().split('\t')
        assert seq not in offsets
        offsets[seq] = (ref, int(offset))

    utils.close(f)
    return offsets


def fix_blast_coords(blast_file, coords_file, outfile):
    coords_offset = offset_coords_file_to_dict(coords_file)
    fin = utils.open_file_read(blast_file)
    fout = utils.open_file_write(outfile)
    for line in fin:
        # blastn sticks a bunch of header lines in the tabulated
        # output file. Need to ignore them
        if '\t' not in line:
            continue

        # Lines are supposed to be tab delimited. Sometimes they
        # have a space character following a tab character, so
        # split on whitespace. This is OK because the pipeline has already
        # removed whitespace from sequence names
        data = line.rstrip().split()
        if data[0] in coords_offset:
            data[6] = str(int(data[6]) + coords_offset[data[0]][1])
            data[7] = str(int(data[7]) + coords_offset[data[0]][1])
            data[0] = coords_offset[data[0]][0]

        # always reconstruct the line, because of spaces bug mentioned above
        line = '\t'.join(data)

        print(line.rstrip(),file=fout)

    utils.close(fin)
    utils.close(fout)

