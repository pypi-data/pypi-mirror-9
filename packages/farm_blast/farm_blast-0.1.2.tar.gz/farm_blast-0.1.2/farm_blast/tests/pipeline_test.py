#!/usr/bin/env python3

import sys
import os
import filecmp
import subprocess
import shutil
import unittest
from farm_blast import pipeline

modules_dir = os.path.dirname(os.path.abspath(pipeline.__file__))
data_dir = os.path.join(modules_dir, 'tests', 'data')

class TestPipeline(unittest.TestCase):
    def setUp(self):
        self.ref = os.path.join(data_dir, 'pipeline_test.ref.fa')
        self.qry = os.path.join(data_dir, 'pipeline_test.qry.fa')
        options = pipeline.get_opts(args=[
            '--no_bsub',
            '--split_bases_tolerance', '1',
            '--split_bases', '100',
            '--outdir', 'tmp.Farm_blast_test',
            '--bsub_name_prefix', 'name',
            '--test',
            self.ref,
            self.qry])
        self.p = pipeline.Pipeline(options, os.path.abspath('scripts/farm_blast'))


    def test_make_setup_script_ref_already_indexed(self):
        expected_script = 'tmp.make_setup_script_expected'
        test_script = 'tmp.make_setup_script_test'
        ref = os.path.join(data_dir, 'pipeline_test.ref.indexed.fa')
        options = pipeline.get_opts(args=[
            '--no_bsub',
            '--split_bases_tolerance', '1',
            '--split_bases', '100',
            '--outdir', 'tmp.Farm_blast_test',
            '--bsub_name_prefix', 'name',
            ref,
            self.qry])
        self.p = pipeline.Pipeline(options, os.path.abspath('scripts/farm_blast'))
        self.p._make_setup_script(script_name=test_script)

        f = open(expected_script, 'w')
        print('set -e', file=f)
        print('fastaq to_fasta -s', os.path.abspath(self.qry), '- |',
              'fastaq chunker --skip_all_Ns - query.split 100 1', file=f)
        f.close()
        self.assertTrue(filecmp.cmp(expected_script, test_script))
        os.unlink(expected_script)
        os.unlink(test_script)


    def test_make_setup_script_convert_ref(self):
        expected_script = 'tmp.make_setup_script_expected'
        test_script = 'tmp.make_setup_script_test'
        self.p._make_setup_script(script_name=test_script)

        f = open(expected_script, 'w')
        print('set -e', file=f)
        print('fastaq to_fasta -s', os.path.abspath(self.ref), 'reference.fa', file=f)
        print('makeblastdb -dbtype nucl -in reference.fa', file=f)
        print('fastaq to_fasta -s', os.path.abspath(self.qry), '- |',
              'fastaq chunker --skip_all_Ns - query.split 100 1', file=f)
        f.close()
        self.assertTrue(filecmp.cmp(expected_script, test_script))
        os.unlink(expected_script)
        os.unlink(test_script)


    def test_make_setup_script_with_union_for_act(self):
        expected_script = 'tmp.make_setup_script_expected'
        test_script = 'tmp.make_setup_script_test'
        self.p.union_for_act = True
        self.p._make_setup_script(script_name=test_script)

        f = open(expected_script, 'w')
        print('set -e', file=f)
        print('fastaq merge', os.path.abspath(self.ref), '- |',
              'fastaq to_fasta -s - reference.fa', file=f)
        print('makeblastdb -dbtype nucl -in reference.fa', file=f)
        print('fastaq merge', os.path.abspath(self.qry), '- |',
              'fastaq to_fasta -s - - |',
              'fastaq chunker --skip_all_Ns - query.split 100 1', file=f)
        f.close()
        self.assertTrue(filecmp.cmp(expected_script, test_script))
        os.unlink(expected_script)
        os.unlink(test_script)


    def test_make_setup_job(self):
        self.p._make_setup_job()
        # the bsub call has a check for home directory, so check everthing after this is ok
        expected = r'''-R "select[mem>1000] rusage[mem=1000]" -M1000 -o 01.setup.sh.o -e 01.setup.sh.e -J name.setup bash 01.setup.sh'''
        self.assertTrue(str(self.p.setup_job).endswith(expected))


    def test_make_start_array_job(self):
        self.p._make_start_array_job()
        # the bsub call has a check for home directory, so check everthing after this is ok
        expected = r'''-o 02.run_array.sh.o -e 02.run_array.sh.e -J name.start_array bash 02.run_array.sh'''
        self.assertTrue(str(self.p.start_array_job).endswith(expected))


    def test_make_array_job(self):
        self.p._make_array_job()
        # the bsub call has a check for home directory, so check everthing after this is ok
        expected = r'''-R "select[mem>500] rusage[mem=500]" -M500 -o tmp.array.o.%I -e tmp.array.e.%I -J "name.array[1-$n]%100" blastn -task blastn -db ''' + self.ref + ''' -query query.split.\$LSB_JOBINDEX -out tmp.array.out.\$LSB_JOBINDEX -outfmt 6'''
        self.assertTrue(str(self.p.array_job).endswith(expected))


    def test_make_start_array_script(self):
        expected_script = os.path.join(data_dir, 'pipeline_test.02.array.sh')
        test_script = 'tmp.make_start_array_script_test'
        self.p.array_job = 'array_job'
        self.p._make_start_array_script(script_name=test_script)
        self.assertTrue(filecmp.cmp(expected_script, test_script))
        os.unlink(test_script)


    def test_make_combine_job(self):
        self.p._make_combine_job()
        # the bsub call has a check for home directory, so check everthing after this is ok
        expected = r'''-n 2 -R "span[hosts=1] select[mem>500] rusage[mem=500]" -M500 -o 03.combine.sh.o -e 03.combine.sh.e -J name.combine bash 03.combine.sh'''
        self.assertTrue(str(self.p.combine_job).endswith(expected))


    def test_make_combine_script(self):
        expected_script = os.path.join(data_dir, 'pipeline_test.03.combine.sh')
        test_script = 'tmp.make_combine_array_script_test'
        self.p._make_combine_script(script_name=test_script)
        expected = [
            'cat tmp.array.e.* > 02.array.e',
            'cat tmp.array.o.* > 02.array.o',
            'cat tmp.array.out.* | gzip -9 -c > blast.out.tmp.gz',
            'PYTHONPATH=',
            '--fix_coords_in_blast_output x x',
            'rm tmp.array.* query.split.* blast.out.tmp.gz 02.array.id 03.combine.sh.id',
            'touch FINISHED'
        ]

        try:
            f = open(test_script)
        except:
            sys.exit('Error opening file', test_script)

        got = [x.rstrip() for x in f.readlines()]
        f.close()

        #Â the script has the absolute path to farm_blast in it. We don't know
        # where that will be, so ignore that part
        self.assertEqual(len(got), len(expected))
        self.assertTrue(got[3].startswith('PYTHONPATH'))
        self.assertTrue('farm_blast' in got[4])
        got[4] = got[4].split(None,1)[1]
        self.assertListEqual(got[:2], expected[:2])
        self.assertListEqual(got[5:], expected[5:])
        os.unlink(test_script)


    def test_pipeline(self):
        self.p.run()
        expected = os.path.join(data_dir, 'pipeline_test.blast.out')
        subprocess.call('gunzip ' + os.path.join(self.p.outdir, 'blast.out.gz'), shell=True)
        self.assertTrue(filecmp.cmp(expected, os.path.join(self.p.outdir, 'blast.out'), shallow=False))


    def tearDown(self):
        if os.path.exists(self.p.outdir):
            shutil.rmtree(self.p.outdir)
