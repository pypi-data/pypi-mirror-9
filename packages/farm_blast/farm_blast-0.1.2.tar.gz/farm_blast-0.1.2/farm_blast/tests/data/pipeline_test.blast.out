qry1	ref1	100.00	100	0	0	1	100	901	1000	4e-50	181
qry2	ref2	100.00	100	0	0	1	100	1	100	4e-50	181
qry2	ref2	100.00	20	0	0	101	120	101	120	1e-07	37.4
qry3	ref2	100.00	19	0	0	102	120	102	120	4e-07	35.6
